test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +279,698 @@ func TestAdoption(t *testing.T) {
+	waitRSStable(t, c, rs)
+}
+
+func updatePod(t *testing.T, podClient typedv1.PodInterface, pod *v1.Pod, updateFunc func(*v1.Pod)) {
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		updateFunc(newPod)
+		_, err = podClient.Update(newPod)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to update pod %s: %v", pod.Name, err)
 	}
-	close(stopCh)
 }
 
-func TestUpdateSelectorToRemoveControllerRef(t *testing.T) {
-	// We have pod1, pod2 and rs. rs.spec.replicas=2. At first rs.Selector
-	// matches pod1 and pod2; change the selector to match only pod1. Verify
-	// that rs creates one more pod, so there are 3 pods. Also verify that
-	// pod2's controllerRef is cleared.
-	s, closeFn, rm, informers, clientSet := rmSetup(t)
+func getPods(t *testing.T, podClient typedv1.PodInterface, labelMap map[string]string) *v1.PodList {
+	podSelector := labels.Set(labelMap).AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	return pods
+}
+
+func TestReplicaSetBasics(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
 	defer closeFn()
-	podInformer := informers.Core().V1().Pods().Informer()
-	ns := framework.CreateTestingNamespace("rs-update-selector-to-remove-controllerref", s, t)
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
 	rs := newRS("rs", ns.Name, 2)
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod1.Labels["uniqueKey"] = "1"
-	pod2 := newMatchingPod("pod2", ns.Name)
-	pod2.Labels["uniqueKey"] = "2"
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	deletingPod := newMatchingPod("deleting-pod", ns.Name)
+	deletingPod.Labels = labelMap
+	deletingPod.DeletionTimestamp = &metav1.Time{Time: time.Now().Add(10 * time.Minute)}
+	deletingPod.Finalizers = []string{"foregroundDeletion"}
+	failedPod := newMatchingPod("failed-pod", ns.Name)
+	failedPod.Labels = labelMap
+	failedPod.Status.Phase = v1.PodFailed
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{deletingPod, failedPod})
+	rs = rss[0]
 
 	stopCh := make(chan struct{})
 	informers.Start(stopCh)
-	waitToObservePods(t, podInformer, 2)
 	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
+	defer close(stopCh)
+	waitRSStable(t, c, rs)
+
+	// First check both deleting and failed pods survive
+	podClient := c.Core().Pods(ns.Name)
+	pods := getPods(t, podClient, labelMap)
+	if len(pods.Items) != 2 {
```

Comment:
If the deleting and failed Pods survived, we should expect 4 total Pods, since RS should ignore them.

Reply:
Yup!
(1) deleting pod
(2) failed pod
(3) new pod replacing the deleting pod
(4) new pod replacing the failed pod